### PR TITLE
✨ Add EIP-6909

### LINKED ISSUE #384: ERC-6909

### PR DESCRIPTION
**Pull request by jtriley-eth**
## Description

Closes #384 

Adds minimal ERC-6909 + tests.

## Checklist

- [x] Ran `forge snapshot`?
- [x] Ran `npm run lint`?
- [x] Ran `forge test`?

### CODE REVIEW

File: src/tokens/ERC6909.sol [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+    function transfer(
+        address receiver,
+        uint256 id,
+        uint256 amount
+    ) public virtual returns (bool) {
+        balanceOf[msg.sender][id] -= amount;
+        balanceOf[receiver][id] += amount;
+        emit Transfer(msg.sender, msg.sender, receiver, id, amount);
+        return true;
```
Comment (transmissions11) [inline src/tokens/ERC6909.sol:41]:
```suggestion
        balanceOf[msg.sender][id] -= amount;
        balanceOf[receiver][id] += amount;

        emit Transfer(msg.sender, msg.sender, receiver, id, amount);

        return true;
```

File: src/tokens/ERC6909.sol [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        uint256 amount
+    ) public virtual returns (bool) {
+        if (msg.sender != sender && !isOperator[sender][msg.sender]) {
+            uint256 allowed = allowance[sender][msg.sender][id];
+            if (allowed != type(uint256).max) allowance[sender][msg.sender][id] = allowed - amount;
+        }
+        balanceOf[sender][id] -= amount;
+        balanceOf[receiver][id] += amount;
+        emit Transfer(msg.sender, sender, receiver, id, amount);
+        return true;
```
Comment (transmissions11) [inline src/tokens/ERC6909.sol:57]:
```suggestion
        if (msg.sender != sender && !isOperator[sender][msg.sender]) {
            uint256 allowed = allowance[sender][msg.sender][id];
            if (allowed != type(uint256).max) allowance[sender][msg.sender][id] = allowed - amount;
        }
        
        balanceOf[sender][id] -= amount;
        balanceOf[receiver][id] += amount;
        
        emit Transfer(msg.sender, sender, receiver, id, amount);
        return true;
```

File: src/tokens/ERC6909.sol [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    function approve(
+        address spender,
+        uint256 id,
+        uint256 amount
+    ) public virtual returns (bool) {
+        allowance[msg.sender][spender][id] = amount;
+        emit Approval(msg.sender, spender, id, amount);
+        return true;
```
Comment (transmissions11) [inline src/tokens/ERC6909.sol:67]:
```suggestion
        allowance[msg.sender][spender][id] = amount;

        emit Approval(msg.sender, spender, id, amount);

        return true;
```

File: src/tokens/ERC6909.sol [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    function setOperator(address operator, bool approved) public virtual {
+        isOperator[msg.sender][operator] = approved;
+        emit OperatorSet(msg.sender, operator, approved);
```
Comment (transmissions11) [inline src/tokens/ERC6909.sol:72]:
```suggestion
        isOperator[msg.sender][operator] = approved;

        emit OperatorSet(msg.sender, operator, approved);
```

File: src/tokens/ERC6909.sol [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+                        INTERNAL MINT/BURN LOGIC
+    //////////////////////////////////////////////////////////////*/
+
+    function _mint(
+        address receiver,
+        uint256 id,
+        uint256 amount
+    ) internal virtual {
+        balanceOf[receiver][id] += amount;
+        totalSupply[id] += amount;
```
Comment (transmissions11) [inline src/tokens/ERC6909.sol:95]: can this not be unchecked?

File: src/tokens/ERC6909.sol [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+                             ERC6909 STORAGE
+    //////////////////////////////////////////////////////////////*/
+
+    mapping(uint256 => uint256) public totalSupply;
+
+    mapping(address => mapping(uint256 => uint256)) public balanceOf;
+
+    mapping(address => mapping(address => mapping(uint256 => uint256))) public allowance;
+
+    mapping(address => mapping(address => bool)) public isOperator;
```
Comment (transmissions11) [inline src/tokens/ERC6909.sol:27]:
```suggestion
    mapping(uint256 => uint256) public totalSupply;

    mapping(address => mapping(address => bool)) public isOperator;

    mapping(address => mapping(uint256 => uint256)) public balanceOf;

    mapping(address => mapping(address => mapping(uint256 => uint256))) public allowance;
```

Comment (transmissions11) [reply src/tokens/ERC6909.sol:27]: for aesthetics :p

File: src/tokens/ERC6909.sol [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+abstract contract ERC6909 {
+    /*//////////////////////////////////////////////////////////////
+                                 EVENTS
+    //////////////////////////////////////////////////////////////*/
+
+    event Transfer(address caller, address indexed from, address indexed to, uint256 indexed id, uint256 amount);
+
+    event OperatorSet(address indexed owner, address indexed operator, bool approved);
+
+    event Approval(address indexed owner, address indexed spender, uint256 indexed id, uint256 amount);
```
Comment (transmissions11) [inline src/tokens/ERC6909.sol:15]:
```suggestion
    event OperatorSet(address indexed owner, address indexed operator, bool approved);

    event Approval(address indexed owner, address indexed spender, uint256 indexed id, uint256 amount);
    
    event Transfer(address caller, address indexed from, address indexed to, uint256 indexed id, uint256 amount);
```

File: src/tokens/ERC6909.sol [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    function _burn(
+        address sender,
+        uint256 id,
+        uint256 amount
+    ) internal virtual {
+        balanceOf[sender][id] -= amount;
+        totalSupply[id] -= amount;
+        emit Transfer(msg.sender, sender, address(0), id, amount);
```
Comment (transmissions11) [inline src/tokens/ERC6909.sol:106]:
```suggestion
        balanceOf[sender][id] -= amount;
        totalSupply[id] -= amount;

        emit Transfer(msg.sender, sender, address(0), id, amount);
```

(same thought about unchecked)

File: src/tokens/ERC6909.sol (hunk 5 above)
Comment (transmissions11) [reply src/tokens/ERC6909.sol:95]: (same spacing change requested here too)

**Review by transmissions11 — COMMENTED** [review]
nice job, very few comments.

only other thing would be can you license this as MIT? dont see a point in making new stuff AGPL

File: src/tokens/ERC6909.sol (hunk 5 above)
Comment (jtriley-eth) [reply src/tokens/ERC6909.sol:102]: Fixed. The `receiver` balance increase on `transfer` and `transferFrom` can also be unchecked. Thoughts on unchecking those as well?

Comment (transmissions11) [reply src/tokens/ERC6909.sol]: yup lets uncheck everything you think is safe, just add a note explaining why like in ERC20 (https://github.com/transmissions11/solmate/blob/fadb2e2778adbf01c80275bfb99e5c14969d964b/src/tokens/ERC20.sol#L186-L187)

**Comment by transmissions11** [comment]
this is probably the best PR i've ever received, thank you for making this easy

**Comment by jtriley-eth** [comment]
ty for the responsiveness 🫡

**Pull request merged**